### PR TITLE
Guided Remediation Docs

### PR DESCRIPTION
**Pull request by michaelkedar**
Link: https://michaelkedar.github.io/osv-scanner/experimental/guided-remediation/

Doc page for guided remediation. Will appreciate feedback if things aren't clear or if something's missing.

#352

### CODE REVIEW

**Comment by michaelkedar** [comment]
I can't see a way to make the subpages expanded by default when not on the parent page :/
image

File: docs/guided-remediation.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+</details>
+
+The `osv-scanner fix` subcommand leverages [deps.dev](https://deps.dev) to provide automated and guided remediation of vulnerabilities in your project's dependencies by suggesting upgrades to dependencies.
+
+Currently, only npm `package.json` manifests and `package-lock.json` lockfiles are supported.
+
+{: .note }
+This feature is experimental and may change in future updates.
+
+## Basic Usage: Interactive Mode
```
Comment (oliverchang) [inline docs/guided-remediation.md:32]: Thinking more, we should start with the CLI usage, and then transition to the interactive mode as an alternative for power users. 

Starting with a heavyweight interactive mode might scare off readers.

File: docs/guided-remediation.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
+---
+layout: page
```
Comment (oliverchang) [inline docs/guided-remediation.md:2]: nit: we should move the other experimental features into dedicated sub pages too for the same/consistent discoverability.

File: docs/guided-remediation.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+{: .note }
+
+> The `package-lock.json` file is regenerated by first deleting the existing `package-lock.json` & `node_modules/` directory, then running `npm install --package-lock-only`. To recreate `node_modules/`, you'll need to run `npm ci` separately
+>
+> The `--relock-cmd` flag can be used to change the executed install command.
+
+## Non-Interactive Mode
+
+To enable usage in scripting and automation, we provide a non-interactive mode:
```
Comment (oliverchang) [inline docs/guided-remediation.md:99]: We should include a sample script in the OSV-Scanner repo that we can link to which shows the automation. We can use the one we wrote before.

File: docs/guided-remediation.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+The relaxation patches are presented in order of effectiveness, with patches that resolve the most vulnerabilities with the least amount of dependency change shown first.
+
+If you wish to apply your current relock & relaxation changes, select the "Write" option to update your manifest file with the new requirements and regenerate your lockfile (if provided).
+
+{: .warning }
+The subcommand cannot revert the changes it makes to your manifest and lockfile. Make sure you backup your files beforehand.
+
+{: .note }
+
+> The `package-lock.json` file is regenerated by first deleting the existing `package-lock.json` & `node_modules/` directory, then running `npm install --package-lock-only`. To recreate `node_modules/`, you'll need to run `npm ci` separately
```
Comment (oliverchang) [inline docs/guided-remediation.md:93]: This bit feels a bit confusing to me. The first sentence does not make it clear that node_modules/ will be NOT be recreated along with package-lock.json. We should clarify that.

**Review by oliverchang — COMMENTED** [review]
nice! some initial comments.

File: docs/guided-remediation.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+  For example, `--min-severity=7.5` will consider only vulnerabilities with CVSS a score of 7.5 and above. If the OSV record for the vulnerability does not have a CVSS score, it will **not** be excluded.
+
+- `--ignore-dev`: Whether to exclude vulnerabilities from dependencies that are only used in `devDependencies`.
+- `--ignore-vulns=<comma-separated list of IDs>`: A list of OSV IDs to exclude from consideration.
+- `--vulns=<comma-separated list of IDs>`: An explicit list of OSV IDs that should exclusively be considered, as long as the other criteria are satisfied.
+
+A vulnerability is only considered if it satisfies all the conditions set by these flags.
+
+### Dependency Upgrade Options
```
Comment (cuixq) [inline docs/guided-remediation.md:194]: do we still consider to refactor this part of flags to only one flag?

**Review by cuixq — APPROVED** [review]

File: docs/guided-remediation.md (hunk 5 above)
Comment (michaelkedar) [reply docs/guided-remediation.md]: I would like to, but I need to implement it first :slightly_smiling_face:

File: docs/guided-remediation.md (hunk 4 above)
Comment (michaelkedar) [reply docs/guided-remediation.md]: I've changed it to be more explicit, but I'm not sure if I've made it any less confusing.

**Review by another-rex — COMMENTED** [review]
Thanks! Added some comments

File: docs/guided-remediation.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+- TOC
+{:toc}
+</details>
+
+The `osv-scanner fix` subcommand leverages [deps.dev](https://deps.dev) to provide automated and guided remediation of vulnerabilities in your project's dependencies by suggesting upgrades to dependencies.
+
+Currently, only npm `package.json` manifests and `package-lock.json` lockfiles are supported.
+
+{: .note }
+This feature is experimental and may change in future updates.
```
[the remark this answers is not in the document]
Comment (michaelkedar) [reply docs/guided-remediation.md]: Changed to use same disclaimer as other experimental bits.

File: docs/guided-remediation.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+![Screenshot of the initial interactive results screen, showing found vulnerabilities and the potential remediation approaches](images/guided-remediation-result.png)
+
+{: .highlight }
+
+> The initial screen allows you to change the vulnerability selection criteria. Each correspond to a [vulnerability selection flag](#vulnerability-selection):
+>
+> - Max dependency depth: `--max-depth`
+> - Min CVSS score: `--min-severity`
+> - Include dev only: `--ignore-dev`
```
Comment (michaelkedar) [inline docs/guided-remediation.md:52]: Surfacing @laserlemon's [comment](https://github.com/michaelkedar/osv-scanner/commit/3def3fad6b4baf7291fd81def72794d307e0a579#r139160741):
> First, thank you for writing this up! 💛
> 
> Is this option really function to "include dev only"? It seems that `--ignore-dev` would mean to only consider _runtime_ dependencies.

You're right that these are inverted of each other - setting `--ignore-dev` would result in `Include dev only: NO`.

I think I'm going to change the interactive screen to 'Exclude dev only' to make them consistent with eachother.

File: docs/guided-remediation.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+Selecting the "Modify lockfile in place" option will bring you to the in-place information page. From here, you can see which vulnerabilities can and cannot be resolved by this strategy. By default, every possible in-place patch will chosen to be applied. You may instead choose which subset of patches you wish to apply.
+
+![Screenshot of the interactive in-place results screen](images/guided-remediation-in-place-results.png)
+
+![Screenshot of the interactive in-place patch selection screen](images/guided-remediation-in-place-choose.png)
+
+If you wish to apply the proposed in-place patches, select the "Write" option to update your lockfile with the new dependency versions.
+
+{: .warning }
```
[the remark this answers is not in the document]
Comment (michaelkedar) [reply docs/guided-remediation.md]: Moved the overwrite/backup warning to basic (non-interactive) section and start of interactive section, just to be clear that both modes will change files.

File: docs/guided-remediation.md (hunk 3 above)
Comment (michaelkedar) [reply docs/guided-remediation.md]: Added a section on scripted usage and linked (and updated) our python script.

File: docs/guided-remediation.md (hunk 1 above)
Comment (michaelkedar) [reply docs/guided-remediation.md]: Moved non-interactive to be 'basic usage', with some forward links to the sections describing in-place/relock strategies.

File: docs/guided-remediation.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+To run it, you can use the following command:
+
+```bash
+osv-scanner fix -M path/to/package.json -L path/to/package-lock.json
+```
+
+The command will launch the Guided Remediation TUI and begin scanning your manifest and lockfile for vulnerabilities. Use the arrow keys to navigate menu items and the `ENTER` key to select. `TAB` can be used to switch to and from the information panel of each option.
+
+![Screenshot of the initial interactive results screen, showing found vulnerabilities and the potential remediation approaches](images/guided-remediation-result.png)
```
[the remark this answers is not in the document]
Comment (michaelkedar) [reply docs/guided-remediation.md:138]: asciinema is neat, but it'll take too much effort to record/edit the files, and hack-ily insert it into just-the-docs.

I might skip recording a gif for now, just so I don't have to keep re-recording it if/when I make changes to the TUI...

**Review by another-rex — APPROVED** [review]
LGTM!

File: docs/guided-remediation.md [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+<details open markdown="block">
+  <summary>
+    Table of contents
+  </summary>
+  {: .text-delta }
+- TOC
+{:toc}
+</details>
+
+The `osv-scanner fix` subcommand leverages [deps.dev](https://deps.dev) to provide automated and guided remediation of vulnerabilities in your project's dependencies by suggesting upgrades to dependencies.
```
Comment (oliverchang) [inline docs/guided-remediation.md:25]: I'd add a paragraph before this to describe the motivation of this and to define "guided remediation".

Something along the lines of:

```
Guided remediation aims to help developers with fixing with the high number of known vulnerabilities in dependencies typically reported by vulnerability scanners by providing a small number of actionable steps. 

This tool provides several options to users for how to prioritise and remediate their vulnerabilities, with easy to understand information on how we arrived at these options and how to compare them. This includes features such as:

- Resolution and analysis of the entire transitive graph to determine the minimal changes required to remove vulnerabilities.
- Prioritising direct dependency upgrades by the total number of transitive vulnerabilities fixed. 
- Prioritising vulnerabilities by dependency depth, severity, and whether or not to care about dev-only dependencies.
- Applying different strategies with different risk/reward ratios (e.g. in-place fixes vs relocking).
```

WDYT?

File: docs/guided-remediation.md [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+![Screenshot of the interactive in-place patch selection screen](images/guided-remediation-in-place-choose.png)
+
+If you wish to apply the proposed in-place patches, select the "Write" option to update your lockfile with the new dependency versions.
+
+{: .note }
+Writing these changes will not reinstall your dependencies. You'll need to run `npm ci` (or equivalent) separately.
+
+## Relock and relax direct dependency remediation
+
+Relocking recomputes your entire dependency graph based on your manifest file, taking the newest possible versions of all your required packages. This causes a large number of changes to your dependency graph, which potentially carries a larger risk of breakages.
```
Comment (oliverchang) [inline docs/guided-remediation.md:167]: Also clarify that these upgrades often unblock constraints allowing more vulnerabilities to be fixed.

File: docs/guided-remediation.md [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+## Known issues
+
+- The subcommand does not use the `osv-scanner.toml` configuration. Use the `--ignore-vulns` flag instead.
+
+### npm
+
+- Non-registry dependencies (local paths, URLs, Git, etc.) are not evaluated.
+- `peerDependencies` are not properly considered during dependency resolution (treated as if using `--legacy-peer-deps`).
+- `overrides` are ignored during dependency resolution.
+- The `node_modules/` in workspaces are not deleted when relocking, which may impact the resulting dependency graph.
```
Comment (oliverchang) [inline docs/guided-remediation.md:247]: Also, should we add a section specifically to call out workspaces and how we handle that?

File: docs/guided-remediation.md (hunk 10 above)
Comment (michaelkedar) [reply docs/guided-remediation.md:27]: I think this is good - I've interspersed it with the existing paragraph.

File: docs/guided-remediation.md (hunk 11 above)
Comment (michaelkedar) [reply docs/guided-remediation.md]: Added another sentence clarifying

File: docs/guided-remediation.md (hunk 12 above)
Comment (michaelkedar) [reply docs/guided-remediation.md]: Expanded upon the workspaces in its own sub-sub-sub-heading :)

File: docs/guided-remediation.md [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+    Table of contents
+  </summary>
+  {: .text-delta }
+- TOC
+{:toc}
+</details>
+
+Guided remediation aims to help developers with fixing with the high number of known vulnerabilities in dependencies typically reported by vulnerability scanners by providing a small number of actionable steps.
+
+The `osv-scanner fix` subcommand leverages [deps.dev](https://deps.dev) to provide automated and guided remediation of vulnerabilities in your project's dependencies by suggesting upgrades to dependencies.
```
Comment (oliverchang) [inline docs/guided-remediation.md:27]: This now feels a bit repetitive. How about folding this into the previous paragraph, and moving the deps.dev mention to the resolution point below?

```
Guided remediation (`osv-scanner fix`) aims to help developers with ....
...

...
- Resolution and analysis of the entire transitive graph (leveraging deps.dev)...
```

**Review by oliverchang — APPROVED** [review]

File: docs/guided-remediation.md (hunk 13 above)
Comment (michaelkedar) [reply docs/guided-remediation.md]: Done

**Review by oliverchang — APPROVED** [review]

**Pull request merged**